src/main/java/org/apache/cassandra/sidecar/cdc/CDCReaderService.java

```
@@ -0,0 +1,157 @@
+                            tableMetadata.params.cdc ? "true" : "false");
+                }
+            }
+            // Start monitoring the cdc_raw directory
+            this.cdcRawDirectoryMonitor.startMonitoring();
+            // Start reading the current commit log.
+            this.cdcIndexWatcher.run();
+
+        }
+        catch (Exception ex)
+        {
+            logger.error("Error starting the CDC reader {}", ex);
+            this.stop();
+            return;
+        }
+        logger.info("Successfully started the CDC reader");
+
+    }
+
+    public synchronized void stop()
+    {
+        logger.info("Stopping CDC reader...");
+        this.cdcRawDirectoryMonitor.stop();
+        this.cdcIndexWatcher.stop();
+        logger.info("Successfully stopped the CDC reader");
+    }
+    @Override
+    public void onAdd(Host host)
+    {
+
+    }
+
+    @Override
+    public void onUp(Host host)
+    {
+
+    }
+
+    @Override
+    public void onDown(Host host)
+    {
+
+    }
+
+    @Override
+    public void onRemove(Host host)
+    {
+
+    }
+
+    @Override
+    public void onRegister(Cluster cluster)
+    {
+
+    }
+
+    @Override
+    public void onUnregister(Cluster cluster)
+    {
+
+    }
+
+    /**
+     * Waiting for the Cassandra server.
+     * */
+    private void waitForCassandraServer() throws InterruptedException
+    {
+        long retryIntervalMs = 1;
+        Cluster cluster = null;
+
+        while (cluster == null)
+        {
+            if (this.session.getLocalCql() != null)
+            {
+                cluster = session.getLocalCql().getCluster();
+            }
+            if (cluster != null)
+            {
+                break;
+            }
+            else
+            {
+                logger.info("Waiting for Cassandra server to start. Retrying after {} milliseconds",
+                        retryIntervalMs);
+                Thread.sleep(retryIntervalMs);
+                retryIntervalMs *= 2;
```

do you think we should add some retry limit ? in this case ,the program will always try to connect to cassandra. 
I think we can add a default retry time such as 3, and init retryIntervalMs can be 10ms, we can retry 3 time ,if cluster is still null we can throw an exception?

---

I modify the code :
long retryIntervalMs = 10;
        int defaultRetryTime = 5;
        Cluster cluster = null;

        while (cluster == null)
        {
            int retryTime = 0;
            if (this.session.getLocalCql() != null)
            {
                cluster = session.getLocalCql().getCluster();
            }
            if (cluster != null)
            {
                break;
            }
            else
            {
                logger.warn("Waiting for Cassandra server to start. Retrying after {} milliseconds",
                    retryIntervalMs);
                if (retryTime++ >= defaultRetryTime)
                    throw new InterruptedException(String.format("Can not connect to cassandra after retry  %s times", defaultRetryTime));
                Thread.sleep(retryIntervalMs);
                retryIntervalMs *= 2;
            }

---

Not sure whether a retry limit is helpful. If the sidecar has to do something useful, it has to wait until the Cassandra starts. If this throws an exception after certain retries, does that mean the sidecar has to stop? or other services are started but not the CDC reader?

---

what I mean is that set the process just sleep round and round is not a good choice, or we can just throw exception after some retry . So the users can go to check what is wrong with the process. Users may not know what is wrong with the process if the deamon is just sleep , they may think that the process is running healthy but actually the cassandra daemon is something wrong and the cdc reader is just sleep.

---

we should let the user know what is going on